controllers/openstackmachine_controller.go

```diff
@@ -292,11 +302,155 @@ func (r *OpenStackMachineReconciler) reconcileDelete(scope scope.Scope, cluster
 		return ctrl.Result{}, fmt.Errorf("delete instance: %w", err)
 	}
 
+	if err := r.reconcileDeleteFloatingAddressFromPool(scope, openStackMachine); err != nil {
+		return ctrl.Result{}, err
+	}
+
 	controllerutil.RemoveFinalizer(openStackMachine, infrav1.MachineFinalizer)
 	scope.Logger().Info("Reconciled Machine delete successfully")
 	return ctrl.Result{}, nil
 }
 
+func (r *OpenStackMachineReconciler) reconcileFloatingIPAddressFromPool(ctx context.Context, scope scope.Scope, openStackMachine *infrav1.OpenStackMachine, openStackCluster *infrav1.OpenStackCluster) (*ipamv1.IPAddressClaim, *ctrl.Result, error) {
+	if openStackMachine.Spec.FloatingAddressFromPool == nil {
+		return nil, nil, nil
+	}
+	var claim *ipamv1.IPAddressClaim
+	claim, err := r.getOrCreateIPAddressClaimForFloatingAddress(ctx, scope, openStackMachine, openStackCluster)
```

Comment:
@mdbooth do we want to bother with the "adoption" design here? Like we're [doing for ports](https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/1788).

Reply:
I've skimmed through the PR and I think I understand what you mean by "adoption", something like restoring a machine if it has a FIP but there's no IPAddressClaim or IPAddress somehow?

Is that a state we can expect to end up in? The current implementation would get a new IP from the pool and most likely fail when trying to associate the new IP unless the IP was restored by tag by the pool and we're lucky enough to get the same IP again.

I guess we could check for a FIP if there's no claim and "adopt" it by creating a claim/ipaddress and that would be picked up by the openstackfloatingippool and handled like any other IP, but it's wouldn't really follow [docs/proposals/20220125-ipam-integration.md](https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20220125-ipam-integration.md) and I don't think we can expect any IPAM provider to pick up any new address from IPAddresses. But that might not be an issue since we expect to get an allocated openstack FIP it's not really like we could use any IPAM provider anyway.

Reply:
Nevermind about "adoption" now. I'm reviewing the API for now. Our goal is to stabilize the API so we can release a v1beta1 in a timely manner.
The rest of it can be iterations in the future.